### PR TITLE
feat(stakingAgent): storage incentives contract calls implementation

### PR DESCRIPTION
**Pull request by swarmHaseeb**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] My change requires updating the Open API specification and/or changing its version, and I've done it.
- [x] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

#### Motivation and context (Optional)

this pr is a continuity of #3347 
### Screenshots (if appropriate):

- - -
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3355)

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -194,25 +195,26 @@ type Options struct {
...
+	Resolver               resolver.Interface
+	Pss                    pss.Interface
+	TraversalService       traversal.Traverser
+	Pinning                pinning.Interface
+	FeedFactory            feeds.Factory
+	Post                   postage.Service
+	PostageContract        postagecontract.Interface
+	Steward                steward.Interface
+	SyncStatus             func() (bool, error)
+	RedistributionContract contract.Interface
```
Comment (istae) [inline pkg/api/api.go:217]: why is this needed ?

Comment (istae) [inline pkg/api/api.go:217]: why is this needed ?

File: pkg/node/node.go [hunk 2]
```diff
@@ -984,26 +985,29 @@ func NewBee(interrupt chan struct{}, sysInterrupt chan os.Signal, addr string, p
...
+		Tags:                   tagService,
+		Storer:                 ns,
+		Resolver:               multiResolver,
+		Pss:                    pssService,
+		TraversalService:       traversalService,
+		Pinning:                pinningService,
+		FeedFactory:            feedFactory,
+		Post:                   post,
+		PostageContract:        postageContractService,
+		RedistributionContract: redistributionService,
```
Comment (istae) [inline pkg/node/node.go:1008]: why is this needed ?

File: pkg/api/api.go (hunk 1 above)
Comment (swarmHaseeb) [reply pkg/api/api.go:217]: because incentives would need this interface to make calls to contract.

Comment (swarmHaseeb) [reply pkg/api/api.go:217]: because incentives would need this interface to make calls to contract.

File: pkg/node/node.go (hunk 2 above)
Comment (swarmHaseeb) [reply pkg/node/node.go:1008]: as above

File: pkg/api/api.go (hunk 1 above)
Comment (swarmHaseeb) [reply pkg/api/api.go:217]: to make it more clear, the IsWinner func within redistribution needs to be called from incentives to actually check if is winner.

File: pkg/incentives/contract/redistribution.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+	ReserveSalt(context.Context) ([]byte, error)
+	IsPlaying(context.Context, uint8) (bool, error)
+	IsWinner(context.Context) (bool, error)
+	Claim(context.Context) error
+	Commit(context.Context, []byte) error
+	Reveal(context.Context, uint8, []byte, []byte) error
+}
+
+type Service struct {
+	overlay                   common.Hash
```
Comment (vladopajic) [inline pkg/incentives/contract/redistribution.go:39]: should `overlay` be `common.Hash` or `swarm.Address`?

File: pkg/incentives/contract/redistribution.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	//TODO: add correct ABI of redistribution contract
+	incentivesContractABI = parseABI(sw3abi.ERC20ABIv0_3_1)
+
+	ErrHasNoStake = errors.New("has no stake")
+)
+
+const loggerName = "redistributionContract"
+
+type Interface interface {
+	ReserveSalt(context.Context) ([]byte, error)
```
Comment (vladopajic) [inline pkg/incentives/contract/redistribution.go:30]: it's not clear what `[]byte` response from `ReserveSalt` represent

File: pkg/incentives/contract/redistribution.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+var (
+	//TODO: add correct ABI of redistribution contract
+	incentivesContractABI = parseABI(sw3abi.ERC20ABIv0_3_1)
+
+	ErrHasNoStake = errors.New("has no stake")
+)
+
+const loggerName = "redistributionContract"
+
+type Interface interface {
```
Comment (vladopajic) [inline pkg/incentives/contract/redistribution.go:29]: definition of `Interface` is not easy to understand, please consider:
 - naming all variables (parameters and response)
 - using appropriate types when possible (`[]byte` is vary generic type)
 - writing some documentation

File: pkg/incentives/contract/redistribution.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+	receipt, err := s.txService.WaitForReceipt(ctx, txHash)
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == 0 {
+		return transaction.ErrTransactionReverted
+	}
+
+	return nil
```
Comment (vladopajic) [inline pkg/incentives/contract/redistribution.go:157]: consider reusing code

following block is repeated in many methods:

```
request := &transaction.TxRequest{
		To:   &s.incentivesContractAddress,
		Data: callData,
	}

	txHash, err := s.txService.Send(ctx, request)
	if err != nil {
		return err
	}

	receipt, err := s.txService.WaitForReceipt(ctx, txHash)
	if err != nil {
		return err
	}

	if receipt.Status == 0 {
		return transaction.ErrTransactionReverted
	}
```

 which could be moved to separate function 
 
 ```
 s.SendAndWait(callData)
 ```

**Review by zelig — COMMENTED** [review]
> Reviewed 3 of 3 files at r1, 4 of 4 files at r2, 1 of 6 files at r4, 1 of 1 files at r5, 1 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3355)* status: 3 of 7 files reviewed, 10 unresolved discussions (waiting on @aloknerurkar, @istae, @swarmHaseeb, @vladopajic, and @zelig)

___
*[`pkg/config/chain.go` line 29 at r2](https://reviewable.io/reviews/ethersphere/bee/3355#-NDDtHnc9un8y1Ff8zcA:-NDDtHnc9un8y1Ff8zcB:b-tihjz3) ([raw file](https://github.com/ethersphere/bee/blob/66f4afedc6aa890f0736b34974f0273d1613ddfb/pkg/config/chain.go#L29)):*
> ```Go
> 	xdaiPostageStampContractAddress   = common.HexToAddress("0xa9c84e9ccC0A0bC9B8C8E948F24E024bC2607c9A")
> 
> 	//Redistibution
> ```

typo missing r
___
*[`pkg/incentives/contract/redistribution.go` line 84 at r2](https://reviewable.io/reviews/ethersphere/bee/3355#-NDDts6qAflkkLhtLR8h:-NDDts6r8gLJdNxjAx8c:b-4bygje) ([raw file](https://github.com/ethersphere/bee/blob/66f4afedc6aa890f0736b34974f0273d1613ddfb/pkg/incentives/contract/redistribution.go#L84)):*
> ```Go
> 
> func (s *Service) IsWinner(ctx context.Context) (isWinner bool, err error) {
> 	//winnerSeed, truthSeed, err := s.checkIsWinning(ctx)
> ```

remove commented code
___
*[`pkg/incentives/contract/redistribution.go` line 207 at r2](https://reviewable.io/reviews/ethersphere/bee/3355#-NDDuBaB6vkWLwk_1-5G:-NDDuBaB6vkWLwk_1-5H:bi7y32g) ([raw file](https://github.com/ethersphere/bee/blob/66f4afedc6aa890f0736b34974f0273d1613ddfb/pkg/incentives/contract/redistribution.go#L207)):*
> ```Go
> }
> 
> // we dont need this for now
> ```

so remove it
___
*[`pkg/incentives/contract/redistribution.go` line 157 at r6](https://reviewable.io/reviews/ethersphere/bee/3355#-ND6P6PTUlpdehnitCyo-r6-157:-NDDtcsK4LJFZbw4jNqV:bi6n0a5) ([raw file](https://github.com/ethersphere/bee/blob/14fc6797f104a9e791c4edc4dd60727c164678a6/pkg/incentives/contract/redistribution.go#L157)):*
<details><summary><i>Previously, vladopajic (Vlado Pajić) wrote…</i></summary><blockquote>

consider reusing code

following block is repeated in many methods:

```
request := &transaction.TxRequest{
		To:   &s.incentivesContractAddress,
		Data: callData,
	}

	txHash, err := s.txService.Send(ctx, request)
	if err != nil {
		return err
	}

	receipt, err := s.txService.WaitForReceipt(ctx, txHash)
	if err != nil {
		return err
	}

	if receipt.Status == 0 {
		return transaction.ErrTransactionReverted
	}
```

 which could be moved to separate function 
 
 ```
 s.SendAndWait(callData)
 ```
</blockquote></details>

agreed

File: pkg/incentives/contract/redistribution.go (hunk 5 above)
Comment (swarmHaseeb) [reply pkg/incentives/contract/redistribution.go:29]: this change was based on a comment by @aloknerurkar for naming conventions.

File: pkg/incentives/contract/redistribution.go (hunk 3 above)
Comment (swarmHaseeb) [reply pkg/incentives/contract/redistribution.go:39]: swarm.Address. Thankyou for highlighting

File: pkg/incentives/redistributioncontract/redistribution.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
```
Comment (istae) [inline pkg/incentives/redistributioncontract/redistribution.go:1]: file and folder should be placed in pkg/storageincentives

**Review by istae — CHANGES_REQUESTED** [review]

File: pkg/storageincentives/redistributioncontract/redistribution_test.go [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+	//nonce := common.BytesToHash(make([]byte, 32))
+	txHashDeposited := common.HexToHash("c3a7")
+
+	t.Run("IsPlaying - true", func(t *testing.T) {
+		t.Parallel()
+
+		depth := uint8(10)
+
+		expectedRes := big.NewInt(1)
+		contract := redistributioncontract2.New(owner, log.NewLogger("logger"),
```
Comment (istae) [inline pkg/storageincentives/redistributioncontract/redistribution_test.go:39]: you can use `log.NoOp` for unit tests

File: pkg/storageincentives/redistributioncontract/redistribution.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+func (s *Service) Claim(ctx context.Context) error {
+	callData, err := redistributionContractABI.Pack("claim")
+	if err != nil {
+		return err
+	}
+	request := &transaction.TxRequest{
+		To:          &s.incentivesContractAddress,
+		Data:        callData,
+		GasPrice:    sctx.GetGasPrice(ctx),
+		GasLimit:    sctx.GetGasLimitWithDefault(ctx, 9_000_000),
```
Comment (istae) [inline pkg/storageincentives/redistributioncontract/redistribution.go:105]: can this also bee 3_000_000

**Review by istae — APPROVED** [review]

File: pkg/storageincentives/redistributioncontract/redistribution.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+		txService:                 txService,
+		incentivesContractAddress: incentivesContractAddress,
+	}
+	return s
+}
+
+func (s *Service) IsPlaying(ctx context.Context, depth uint8) (bool, error) {
+	callData, err := redistributionContractABI.Pack("isParticipatingInUpcomingRound", common.BytesToHash(s.overlay.Bytes()), depth)
+	if err != nil {
+		return false, err
```
Comment (aloknerurkar) [inline pkg/storageincentives/redistributioncontract/redistribution.go:61]: Errors from top-level functions should ideally we wrapped with some information about redistribution contract.

I see not debug information coming from this package. You will only see errors from geth which will be hard to debug.

File: pkg/storageincentives/redistributioncontract/redistribution_test.go [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethersphere/bee/pkg/log"
+	redistributioncontract2 "github.com/ethersphere/bee/pkg/storageincentives/redistributioncontract"
```
Comment (aloknerurkar) [inline pkg/storageincentives/redistributioncontract/redistribution_test.go:18]: Why is this named 2?

File: pkg/storageincentives/redistributioncontract/redistribution.go (hunk 9 above)
Comment (swarmHaseeb) [reply pkg/storageincentives/redistributioncontract/redistribution.go:105]: i discussed with Abel today he said that its better to have 9_000_000 for claim as it may use more gas.

**Review by mrekucci — COMMENTED** [review]
The package name is too long and describes what the package contains. Rather, the name should express a concept, idea or behaviour: https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#packages

**Review by metacertain — APPROVED** [review]
LGTM, thanks!

**Review by vladopajic — APPROVED** [review]
LGTM
